src/main/java/net/neoforged/neoforge/common/crafting/NBTIngredient.java

```diff
@@ -0,0 +1,109 @@
+ */
+public class NBTIngredient extends Ingredient {
+
+    public static final Codec<NBTIngredient> CODEC = RecordCodecBuilder.create(
+            builder -> builder
+                    .group(
+                            NeoForgeExtraCodecs.singularOrPluralCodec(BuiltInRegistries.ITEM.byNameCodec(), "item").forGetter(NBTIngredient::getContainedItems),
+                            CraftingHelper.TAG_CODEC.fieldOf("tag").forGetter(NBTIngredient::getTag),
+                            Codec.BOOL.optionalFieldOf("strict", false).forGetter(NBTIngredient::isStrict))
+                    .apply(builder, NBTIngredient::new));
+
+    public static final Codec<NBTIngredient> CODEC_NONEMPTY = RecordCodecBuilder.create(
+            builder -> builder
+                    .group(
+                            NeoForgeExtraCodecs.singularOrPluralCodecNotEmpty(BuiltInRegistries.ITEM.byNameCodec(), "item").forGetter(NBTIngredient::getContainedItems),
+                            CraftingHelper.TAG_CODEC.fieldOf("tag").forGetter(NBTIngredient::getTag),
+                            Codec.BOOL.optionalFieldOf("strict", false).forGetter(NBTIngredient::isStrict))
+                    .apply(builder, NBTIngredient::new));
+
+    private final boolean strict;
+
+    protected NBTIngredient(Set<Item> items, CompoundTag tag, boolean strict) {
+        super(items.stream().map(item -> {
+            ItemStack stack = new ItemStack(item, 1);
+            // copy NBT to prevent the stack from modifying the original, as attachments or vanilla item durability will modify the tag
+            stack.setTag(tag.copy());
+            return new Ingredient.ItemValue(stack, strict ? ItemStack::matches : NBTIngredient::compareStacksWithNBT);
+        }), NeoForgeMod.NBT_INGREDIENT_TYPE);
+        Preconditions.checkArgument(!items.isEmpty(), "At least one item needs to be provided for a nbt matching ingredient");
+        this.strict = strict;
+    }
+
+    @Override
+    protected boolean areStacksEqual(ItemStack left, ItemStack right) {
+        return strict ? ItemStack.matches(left, right) : compareStacksWithNBT(left, right);
+    }
+
+    @Override
+    public boolean synchronizeWithContents() {
+        return false;
+    }
+
+    private static boolean compareStacksWithNBT(ItemStack left, ItemStack right) {
+        return left.getItem() == right.getItem() && NbtUtils.compareNbt(left.getTag(), right.getTag(), true);
+    }
+
+    /**
+     * Creates a new ingredient matching any item from the list, containing the given NBT
+     */
+    public static NBTIngredient of(boolean strict, CompoundTag nbt, ItemLike... items) {
+        return new NBTIngredient(Arrays.stream(items).map(ItemLike::asItem).collect(Collectors.toSet()), nbt, strict);
+    }
+
+    /**
+     * Creates a new ingredient matching the given item, containing the given NBT
+     */
+    public static NBTIngredient of(boolean strict, ItemStack stack) {
```

Comment:
This really needs a javadoc explaining that non-strict matching may give unexpected results when using item stacks as they automatically set certain NBT keys (e.g. damagable items will be forced to match against a damage value of 0)